### PR TITLE
JAVA-2665: Migrate Job-Creator build.yaml to Jenkins Pipeline Jenkinsfile

### PR DESCRIPTION
**Pull request by emerkle826**
This PR addresses a bit more than just [JAVA-2665](https://datastax-oss.atlassian.net/browse/JAVA-2665). It also addresses a few of issues with tests found running against DSE 6.8:

1) The first issue is that when integration tests fail due to CCM failing to start, the error reported may not be very helpful in determining why CCM failed to start. Sometimes there are details that can be obtained by running `ccm checklogerror` that will make it easier to determine the issue. The changes to `CcmBrideg.java` attempt to run that command in those cases.

2) The second issue is that `DefaultReactiveGraphResultIT` is marked as a `ParallelizableTest`, but uses `CustomCcmRule`. This causes 2 attempts to start a CCM instance during the parallel tests, one for the shared `CcmRule` and one for the shared `CustomCcmRule`. Removing the `ParallelizableTest` annotation from this test fixes the issue (and effectively makes it a Serializable test.

3) The third issue is that there is a bug when creating SASI indexes in DSE 6.8.0 [DSP-20720](https://datastax.jira.com/browse/DSP-20720). This causes many of the Mapper ITs to fail when running against DSE 6.8.0, and also causes other ITs to fail that try to use the `CcmRule` after them. I tried to fix this by preventing SASI indexes from being created when running against DSE 6.8.0 and skip any tests that require SASI indexes when running against DSE 6.8.0.

### CODE REVIEW

**Comment by emerkle826** [comment]
<b>NOTE:</b> Once this PR is merged into `4.x`, Jenkins builds for `4.x` will no longer be picked up by Job-Creator and the old Datastax Jenkins build servers.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,599 @@
+#!groovy
+
+def initializeEnvironment() {
+  env.DRIVER_DISPLAY_NAME = 'CassandraⓇ Java Driver'
+  env.DRIVER_METRIC_TYPE = 'oss'
+  if (env.GIT_URL.contains('riptano/java-driver')) {
```
Comment (tomekl007) [inline Jenkinsfile:6]: shouldn't it be `datastax/java-driver`?

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+  '''
+}
+
+def buildDriver(jabbaVersion) {
+  withEnv(["BUILD_JABBA_VERSION=${jabbaVersion}"]) {
+    sh label: 'Build driver', script: '''#!/bin/bash -lex
+      . ${JABBA_SHELL}
+      jabba use ${BUILD_JABBA_VERSION}
+
+      mvn -B -V install -DskipTests -Dmaven.javadoc.skip=true
```
Comment (tomekl007) [inline Jenkinsfile:55]: shouldn't we do `clean install` to make sure that the build will be fresh?

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    env.BUILD_STATED_SLACK_NOTIFIED = 'true'
+
+    def buildType = 'Commit'
+    if (params.CI_SCHEDULE != 'DO-NOT-CHANGE-THIS-SELECTION') {
+      buildType = "${params.CI_SCHEDULE.toLowerCase().capitalize()}"
+    }
+
+    def color = 'good' // Green
+    if (status.equalsIgnoreCase('aborted')) {
+      color = '808080' // Grey
```
Comment (tomekl007) [inline Jenkinsfile:114]: why other colors are named and this is number?

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/InventoryITBase.java [hunk 4]
```diff
@@ -85,6 +86,13 @@
   }
 
   private static final Version MINIMUM_SASI_VERSION = Version.parse("3.4.0");
+  private static final Version BROKEN_SASI_VERSION = Version.parse("6.8.0");
+
+  protected static boolean isSasiBroken(CcmRule ccmRule) {
+    Optional<Version> dseVersion = ccmRule.getDseVersion();
+    // creating SASI indexes is broken in DSE 6.8.0
```
Comment (tomekl007) [inline integration-tests/src/test/java/com/datastax/oss/driver/mapper/InventoryITBase.java:93]: shouldn't we report it somewhere? Is there a JIRA issue that captures that?

**Comment by tomekl007** [comment]
@emerkle826 - where the new Jenkins build can be seen?

**Comment by emerkle826** [comment]
> @emerkle826 - where the new Jenkins build can be seen?

Sorry, I forgot to mention it. It's located here: [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/jov/oss/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/)

Currently, there is a folder (instead of the old Jenkins Views), and the folder has 3 sub-folders, one for the OSS driver repo, one for the DSE driver repo, and one for the private development repo. Those are here: [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/)

File: Jenkinsfile (hunk 1 above)
Comment (emerkle826) [reply Jenkinsfile:6]: no, that is for the Private fork of the OSS driver. This is just setting up a pipeline variable named `DRIVER_DISPLAY_NAME` that is used when notifying the Slack bot room, to help identify whether an event for a build is related to the OSS driver, the DSE driver or the private fork of the OSS driver.

Each of those repositories has its own section in the new Jenkins server:
Folder: [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/)
riptano/java-driver (called Java Development): [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/dev/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/dev/)
riptano/java-dse-driver (called Java (DSE)): [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/dse/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/dse/)
datastax/java-driver (called Java (OSS)): [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/)

File: Jenkinsfile (hunk 2 above)
Comment (emerkle826) [reply Jenkinsfile:55]: We don't need to as this happens on a brand new Openstack instance that has just checked out the branch from GitHub, so there is no previous compilation attempt in the environment when this runs.

File: Jenkinsfile (hunk 3 above)
Comment (emerkle826) [reply Jenkinsfile:114]: This is the Slack plugin for Jenkins. It only understands those 3 named colors, and any hex code color: [https://jenkins.io/doc/pipeline/steps/slack/](https://jenkins.io/doc/pipeline/steps/slack/)

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/InventoryITBase.java (hunk 4 above)
Comment (emerkle826) [reply integration-tests/src/test/java/com/datastax/oss/driver/mapper/InventoryITBase.java:93]: I opened [DSP-20720](https://datastax.jira.com/browse/DSP-20720) against this. It's mentioned in the PR description above too. :wink:

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+  }
+}
+
+def describeAdhocAndScheduledTestingStage() {
+  script {
+    currentBuild.displayName = "Adhoc testing"
+    if (params.CI_SCHEDULE == 'WEEKNIGHTS') {
+      currentBuild.displayName = "${params.CI_SCHEDULE.toLowerCase().replaceAll('_', ' ').capitalize()} schedule"
+    }
+    currentBuild.description = "Testing ${params.ADHOC_BUILD_AND_EXECUTE_TESTS_SERVER_VERSION} against JDK version ${params.ADHOC_BUILD_AND_EXECUTE_TESTS_JABBA_VERSION}"
```
Comment (emerkle826) [inline Jenkinsfile]: This will actually have the wrong information in the job description for scheduled builds. See [https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/30/](https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/30/) as an example. That job description indicates it was run against Cassandra 2.1 with JDK8, when the parameterized build info below shows it was actually run against 3.11, 4.0, 6.7 and 6.8 using JDK 14. I'll fix this in my next commit.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+  triggers {
+    parameterizedCron("""
+      # Every weeknight (Monday - Friday) around 2:00 AM
+      ### JDK8 tests against 2.1, 3.0, DDAC-5.1, DSE 4.8, DSE 5.0, DSE 5.1, DSE-6.0 and DSE 6.7
+      H 2 * * 1-5 %CI_SCHEDULE=WEEKNIGHTS;CI_SCHEDULE_SERVER_VERSIONS=2.1 3.0 ddac-5.1 dse-4.8 dse-5.0 dse-5.1 dse-6.0 dse-6.7;CI_SCHEDULE_JABBA_VERSION=1.8
+      ### JDK11 tests against 3.11, 4.0, DSE 6.7 and DSE 6.8
+      H 2 * * 1-5 %CI_SCHEDULE=WEEKNIGHTS;CI_SCHEDULE_SERVER_VERSIONS=3.11 4.0 dse-6.7 dse-6.8;CI_SCHEDULE_JABBA_VERSION=openjdk@1.11
+      # Every weekend (Sunday) around 12:00 PM noon
+      ### JDK14 tests against 3.11, 4.0, DSE 6.7 and DSE 6.8
+      H 12 * * 0 %CI_SCHEDULE=WEEKNIGHTS;CI_SCHEDULE_SERVER_VERSIONS=3.11 4.0 dse-6.7 dse-6.8;CI_SCHEDULE_JABBA_VERSION=openjdk@1.14
```
Comment (emerkle826) [inline Jenkinsfile]: This probably should use `CI_SCHEDULE=WEEKENDS` instead of `WEEKNIGHTS`.

File: Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+      defaultValue: false,
+      description: 'Flag to determine if parallel integration tests should be skipped')
+    string(
+      name: 'INTEGRATION_TESTS_FILTER',
+      defaultValue: '',
+      description: '''<p>Run only the tests whose name match patterns</p>
+                      See <a href="https://maven.apache.org/surefire/maven-failsafe-plugin/examples/single-test.html">Maven Failsafe Plugin</a> for more information on filtering integration tests''')
+    choice(
+      name: 'CI_SCHEDULE',
+      choices: ['DO-NOT-CHANGE-THIS-SELECTION', 'WEEKNIGHTS'],
```
Comment (emerkle826) [inline Jenkinsfile]: For the future, we should add WEEKENDS and MONTHLY to the list.

**Comment by tomekl007** [comment]
Is it normal that build for this branch fails:
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/
?

**Comment by emerkle826** [comment]
> Is it normal that build for this branch fails:
> https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/
> ?

It is right now as the Per-Commit builds happen against Cassandra 4.0, DSE 6.7 and DSE 6.8.0. Most of the time, the tests pass against 4.0 and 6.7. I'm still working on the 6.8 failures.

**Comment by tomekl007** [comment]
@emerkle826 When did you create this file `Jenkinsfile` is it automatically fetched by the Jenkins Pipeline, or did you need to do some manual step (i.e. registering it in the Jenkins)?

**Comment by emerkle826** [comment]
> @emerkle826 When did you create this file `Jenkinsfile` is it automatically fetched by the Jenkins Pipeline, or did you need to do some manual step (i.e. registering it in the Jenkins)?

It was created from scratch (by Fero). I've been making changes to customize our build process. The repo is setup on the CloudBees Jenkins server, and similar to how Job Creator would find `build.yaml` files in branches, the CloudBees server finds branches with Jenkinsfile files.

To get new GitHub repos into CloudBees, I believe we have to manually setup the repo. I have a doc that is in progress that should hopefully cover that.

**Comment by adutra** [comment]
> The second issue is that `DefaultReactiveGraphResultIT` is marked as a `ParallelizableTest`, but uses `CustomCcmRule`. 

I think you should instead keep that test parallelizable but use the regular `CcmRule` instead. It doesn't do anything fancy with CCM so it is certainly parallelizable.

File: Jenkinsfile (hunk 1 above)
Comment (adutra) [reply Jenkinsfile:6]: Is everybody ok with these names? I thought we might switch to something more obvious, e.g. `datastax-oss`, `riptano-oss`, `riptano-dse`. Just a suggestion, no need to change names right away.

File: Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+  }
+
+  env.GIT_SHA = "${env.GIT_COMMIT.take(7)}"
+  env.GITHUB_PROJECT_URL = "https://${GIT_URL.replaceFirst(/(git@|http:\/\/|https:\/\/)/, '').replace(':', '/').replace('.git', '')}"
+  env.GITHUB_BRANCH_URL = "${GITHUB_PROJECT_URL}/tree/${env.BRANCH_NAME}"
+  env.GITHUB_COMMIT_URL = "${GITHUB_PROJECT_URL}/commit/${env.GIT_COMMIT}"
+
+  env.MAVEN_HOME = "${env.HOME}/.mvn/apache-maven-3.3.9"
+  env.PATH = "${env.MAVEN_HOME}/bin:${env.PATH}"
+  env.JAVA_HOME = sh(label: 'Get JAVA_HOME',script: '''#!/bin/bash -lex
```
Comment (adutra) [inline Jenkinsfile]: Fero said something about removing the -x switch to reduce verbosity, is that planned for later?

File: Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+    . ${JABBA_SHELL}
+    jabba which 1.8''', returnStdout: true).trim()
+
+  sh label: 'Download Apache CassandraⓇ or DataStax Enterprise',script: '''#!/bin/bash -lex
+    . ${JABBA_SHELL}
+    jabba use ${JABBA_VERSION}
+    . ${CCM_ENVIRONMENT_SHELL} ${SERVER_VERSION}
+  '''
+
+  sh label: 'Display Java and environment information',script: '''#!/bin/bash -le
```
Comment (adutra) [inline Jenkinsfile:34]: I confirm that this step in Blue Ocean is much less verbose than the 2 previous ones that are still using -x

File: Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+            steps {
+              initializeEnvironment()
+              script {
+                if (env.BUILD_STATED_SLACK_NOTIFIED != 'true') {
+                  notifySlack()
+                }
+              }
+            }
+          }
+          stage('Describe-Build') {
```
Comment (adutra) [inline Jenkinsfile:432]: So this stage appears in Blue Ocean as having no steps, why is that?

File: Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+                 *
+                 *  - Build failures during mvn verify may exist so report may not be available
+                 */
+                junit testResults: '**/target/surefire-reports/TEST-*.xml', allowEmptyResults: true
+                junit testResults: '**/target/failsafe-reports/TEST-*.xml', allowEmptyResults: true
+              }
+            }
+          }
+          stage('Execute-Code-Coverage') {
+            // Ensure the code coverage is run only once per-commit
```
Comment (adutra) [inline Jenkinsfile:462]: Code coverage, as far as I understand, is being performed for all runs, but we are only collecting and displaying results for SERVER_VERSION = 4.0. Not sure if it's useful to limit the coverage displaying without limiting the coverage performing.
Also, I'm not seeing code coverage results anywhere in Blue Ocean. Not sure if they are being correctly collected.

File: integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java [hunk 12]
```diff
@@ -36,14 +35,12 @@
...
 import org.junit.Test;
-import org.junit.experimental.categories.Category;
 import org.junit.rules.RuleChain;
 import org.junit.rules.TestRule;
 import org.junit.runner.RunWith;
 
 @DseRequirement(min = "6.8.0", description = "Graph paging requires DSE 6.8+")
 @RunWith(DataProviderRunner.class)
-@Category(ParallelizableTests.class)
 public class DefaultReactiveGraphResultSetIT {
```
Comment (adutra) [inline integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java:44]: As mentioned earlier, I think it's better to keep this parallelizable and change the rule below.

Comment (emerkle826) [reply integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java:44]: I didn't even think about trying that, but I will.

File: Jenkinsfile (hunk 1 above)
Comment (emerkle826) [reply Jenkinsfile:6]: We can definitely change the names, discuss it at the next stand-up.

File: Jenkinsfile (hunk 8 above)
Comment (emerkle826) [reply Jenkinsfile]: We can, but the general approach has been that most of the time, you don't even need to look at this stuff. But when you do, it's usually because something is not working as expected, and then it's helpful to have more output (even if it's a large mass to sift through).

File: Jenkinsfile (hunk 10 above)
Comment (emerkle826) [reply Jenkinsfile:432]: it's because the the steps section below only runs [describePerCommitStage](https://github.com/datastax/java-driver/pull/1416/files#diff-58231b16fdee45a03a4ee3cf94a9f2c3R153-R158), which doesn't do anything but change the build name and description. I'll look at the C++ Jenkinsfile to see what is happening in there. We may be able to just eliminate this step.

File: Jenkinsfile (hunk 11 above)
Comment (emerkle826) [reply Jenkinsfile:462]: Code coverage has a few issues with this setup. Since each combination is run on a separate VM, aggregating the results is going to be a challenge. Also, the JaCoCo plugin for CloudBees is on version 3.0.4, which does not collect the coverage metrics and make the charts in Jenkins if any phase of the pipeline reports as failed or unstable. So unless every test run passes, the coverage metrics aren't collected. Version 3.0.5 of the plugin allows for forcing the collection to happen, even if tests fail. That version of the plugin is not yet available via the CloudBees installation, I can look into upgrading it.

The metrics have been collected for a few of the runs that completed with all tests passing, but I think you can only see them on the regular Jenkins UI (i.e. not Blue Ocean). You can see the coverage trend here:
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/jacoco/
And the last run here:
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/40/
and
https://jenkins-drivers.build.dsinternal.org/job/drivers/job/java/job/oss/job/java2665/40/jacoco

TL;DR; The code coverage stuff needs more attention.

File: integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java (hunk 12 above)
Comment (emerkle826) [reply integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java:44]: Actually, I don't think I can switch this to use `CcmRule` as the Rule needs to be able to set the "graph" workload.

Comment (adutra) [reply integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java:44]: Ouch, you are right. We should investigate if it's possible to add the graph workload to the default CCM cluster for parallelizable tests.

File: Jenkinsfile (hunk 11 above)
Comment (adutra) [reply Jenkinsfile:462]: > That version of the plugin is not yet available via the CloudBees installation, I can look into upgrading it.

Thanks for the explanations, would you mind filing a Jira for the code coverage issues and details what you think should/could be improved?

Also, I was looking at the dashboards you linked, the coverage numbers seem excessively low. We might want to check their correctness by comparing with a local build.

File: integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java (hunk 12 above)
Comment (adutra) [reply integration-tests/src/test/java/com/datastax/dse/driver/api/core/graph/reactive/DefaultReactiveGraphResultSetIT.java:44]: Created JAVA-2723.

**Review by adutra — APPROVED** [review]

File: Jenkinsfile (hunk 11 above)
Comment (emerkle826) [reply Jenkinsfile:462]: Jira filed here: https://datastax-oss.atlassian.net/browse/JAVA-2724.
I'll check the coverage we are currently getting with a build/test against 4.0 locally and on CloudBees.

File: Jenkinsfile (hunk 10 above)
Comment (emerkle826) [reply Jenkinsfile:432]: Just to confirm, the C++ (and I believe all the other POC Jenkinsfile files) all do pretty much the same thing. This step is merely a cosmetic one for changing the name and description of the build in the Blue Ocean UI.

File: Jenkinsfile (hunk 11 above)
Comment (adutra) [reply Jenkinsfile:462]: Thank you!
